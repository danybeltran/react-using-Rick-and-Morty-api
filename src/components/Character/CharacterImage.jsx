import React from "react";

export const CharacterImage = ({ image, name }) => {
    return <img className="character-image" src={image} alt={name + " picture"} />
}