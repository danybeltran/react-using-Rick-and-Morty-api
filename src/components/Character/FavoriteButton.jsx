import React, { useState, useEffect, useContext } from "react";
import { Liked, HeartLogo } from "../HeartLogo";
import { AppContext } from "../../Context";

export const FavoriteButton = ({ id }) => {

    if (typeof localStorage["favorite"] == "undefined") {
        localStorage["favorite"] = "[]"
    }

    const context = useContext(AppContext);

    var clicked, setClicked;

    if (typeof context === "undefined") {
        clicked = 0;
        setClicked = function (newValue) {
            clicked++
        }
    }
    else {
        clicked = context.clicked;
        setClicked = context.setClicked;
    }
    // Favorite characters are saved in the localstorage
    var [favoriteList, setFavoriteList] = useState(JSON.parse(localStorage["favorite"]));


    useEffect(() => {
        setFavoriteList(JSON.parse(localStorage["favorite"]));
    }, [clicked])

    let addOrRemoveToFavoriteList = () => {

        if (isFavorite) {
            // Remove the item if it's liked
            favoriteList.splice(favoriteList.indexOf(id), 1);
            localStorage["favorite"] = JSON.stringify(favoriteList);
            setClicked(clicked + 1)
        } else {

            // Max of 5 items
            if (favoriteList.length < 5) {
                favoriteList.push(id);
                localStorage["favorite"] = JSON.stringify(favoriteList);
                setClicked(clicked + 1)
            } else {
                alert("You can only ad up to 5 characters to the 'Favorite' list");
            }
        }

    }

    let isFavorite = favoriteList.includes(id)
    return (
        <button className="favorite-button flex" onClick={addOrRemoveToFavoriteList} title={(isFavorite ? "Remove from" : "Add to") + " favorite"}>
            {/* {// Render a white heart if not favorite and filled if it is} */}
            {isFavorite ?
                <>
                    <Liked />
                    <span className="pl-3 text-sm md:text-lg">
                        Remove
                </span>
                </> :
                <>
                    <HeartLogo />
                    <span className="pl-3 text-sm md:text-lg">
                        Favorite
                </span>
                </>}
        </button>
    )
}