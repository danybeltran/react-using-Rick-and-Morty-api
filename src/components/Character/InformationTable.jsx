import React from "react";

export const InformationTable = ({ name, species, episode, status }) => {

    let CharacterStatus = status === "Alive" ? "text-green-600" :
        status === "Dead" ? "text-red-600" : "text-yellow-600";

    return (
        <div className="informationTable">
            <p className="character-name">{name}</p>
            <p>Species: {species}</p>
            <p>Episodes: {episode.length}</p>
            <p>Status:<span className={CharacterStatus}> {status}</span></p>
        </div>
    )
}