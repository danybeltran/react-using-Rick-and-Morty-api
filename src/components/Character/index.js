export { CharacterImage }
from "./CharacterImage"
export { FavoriteButton }
from "./FavoriteButton"
export { Information }
from "./Information"
export { InformationTable }
from "./InformationTable"
export { Character }
from "../Character";