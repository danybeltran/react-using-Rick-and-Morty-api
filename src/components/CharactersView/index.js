export { CharactersBody }
from "./CharactersBody";
export { CharactersContainer }
from "./CharactersContainer";
export { CharactersHeader }
from "./CharactersHeader";
export { MappedCharacters }
from "./MappedCharacters";
export { Pagination }
from "./Pagination"
export { PaginationLink }
from "./PaginationLink";
export { Results }
from "./Results"
export { SearchButton }
from "./SearchButton";
export { SearchInput }
from "./SearchInput";
export { ZeroFound }
from "./ZeroFound";