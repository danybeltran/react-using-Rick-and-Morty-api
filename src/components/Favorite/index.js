export { FavoriteCharacters } from "./FavoriteCharacters"
export { FavoriteCharactersHeader } from "./FavoriteCharactersHeader"
export { FavoriteContainer } from "./FavoriteContainer"
export { MappedFavoriteCharacters } from "./MappedFavoriteCharacters"
export { NoResults } from "./NoResults"