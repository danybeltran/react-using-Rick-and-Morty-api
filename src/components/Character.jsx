import React from "react";
import { CharacterImage } from "./Character/CharacterImage"
import { FavoriteButton } from "./Character/FavoriteButton";
import { InformationTable } from "./Character/InformationTable";

export const Character = (props, { isFavorite, userWillSearch }) => {

    const { status } = props;

    let CharacterStatus = status === "Alive" ? "text-green-600" :
        status === "Dead" ? "text-red-600" : "text-yellow-600";

    return (
        <>
            <div className="character">
                <div className="character-body">
                    <CharacterImage {...props} />
                    <div className="character-information">
                        <InformationTable {...props} />
                        <br />
                        <FavoriteButton {...{ id: props.id, isFavorite }} />
                    </div>
                </div>
            </div>
        </>
    )
}