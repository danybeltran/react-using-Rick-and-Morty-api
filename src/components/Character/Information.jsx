import React from "react";
import { InformationTable } from "./InformationTable";
import { FavoriteButton } from "./FavoriteButton";

export const Information = (props) => {

    return (
        <div className="character-information">
            <InformationTable {...props} />
            <br />
            <FavoriteButton {...props} />
        </div>
    )
}