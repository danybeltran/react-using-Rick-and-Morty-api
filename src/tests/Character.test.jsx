import React from "react";
import renderer from "react-test-renderer";
import { Character } from "../components/Character/";


test("Should render Rick Sanchez character information", () => {
    const renderdComponent = renderer.create(
        <Character
            id={2}
            image="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
            name="Rick Sanchez"
            episode={["1", "2"]}
            status="Alive"
            species="Human"
            isFavorite={true}
        />
    );

    let component = renderdComponent.toJSON();

    expect(component).toMatchSnapshot();

    expect


})