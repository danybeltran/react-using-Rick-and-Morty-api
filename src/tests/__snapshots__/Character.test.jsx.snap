// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Rick Sanchez character information 1`] = `
<div
  className="character"
>
  <div
    className="character-body"
  >
    <img
      alt="Rick Sanchez picture"
      className="character-image"
      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
    />
    <div
      className="character-information"
    >
      <div
        className="informationTable"
      >
        <p
          className="character-name"
        >
          Rick Sanchez
        </p>
        <p>
          Species: 
          Human
        </p>
        <p>
          Episodes: 
          2
        </p>
        <p>
          Status:
          <span
            className="text-green-600"
          >
             
            Alive
          </span>
        </p>
      </div>
      <br />
      <button
        className="favorite-button flex"
        onClick={[Function]}
        title="Add to favorite"
      >
        <svg
          className="mt-1 bi bi-heart"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 2.748l-.717-.737C5.6.281 2.514.878 1.4 3.053c-.523 1.023-.641 2.5.314 4.385.92 1.815 2.834 3.989 6.286 6.357 3.452-2.368 5.365-4.542 6.286-6.357.955-1.886.838-3.362.314-4.385C13.486.878 10.4.28 8.717 2.01L8 2.748zM8 15C-7.333 4.868 3.279-3.04 7.824 1.143c.06.055.119.112.176.171a3.12 3.12 0 0 1 .176-.17C12.72-3.042 23.333 4.867 8 15z"
            fillRule="evenodd"
          />
        </svg>
        <span
          className="pl-3 text-sm md:text-lg"
        >
          Favorite
        </span>
      </button>
    </div>
  </div>
</div>
`;
